datalad/support/annexrepo.py

```diff
@@ -12,23 +12,53 @@
 
 """
 
-from os.path import join, exists
+from os import getcwd
+from os.path import join as p_join, exists, normpath, isabs, commonprefix, relpath
 import logging
 
 from ConfigParser import NoOptionError
 
-from gitrepo import GitRepo
+from gitrepo import GitRepo, files_decorator
 from datalad.cmd import Runner as Runner
+from exceptions import CommandNotAvailableError, CommandError, FileNotInAnnexError, FileInGitError
 
 lgr = logging.getLogger('datalad.annex')
 
 
+def _options_decorator(func):
+    """Decorator to provide convenient way to pass options to command calls.
+
+    Any keyword argument "foo='bar'" translates to " --foo=bar".
+    All of these are collected in a list and then passed to keyword argument `options`
+    of the decorated function.
+    Note: This is meant to especially decorate the methods of AnnexRepo-class and therefore
+    returns a class method.
+    """
+
+    def newfunc(self, *args, **kwargs):
+        option_list=[]
+        for key in kwargs.keys():
+            option_list.extend([" --%s=%s" % (key, kwargs.get(key))])
+
+        return func(self, *args, options=option_list)
+    return newfunc
+
+
+
 class AnnexRepo(GitRepo):
     """Representation of an git-annex repository.
 
-    """
 
-    def __init__(self, path, url=None, runner=None):
+    Paths given to any of the class methods will be interpreted as relative to os.getcwd() in case this is currently beneath
```

Comment:
datalad commands will have to act like any git/git annex command in the
end, right? So, usually there will be a relative path given via command
line from within or beneath the repository's root. These paths can passed
as they are to the object representing this very repository. If we for some
reason use such an object while running outside it's repository, it will
then take relative paths as being relative to it's root. Absolute paths are
always accepted. To me this seems to be the most intuitiv handling. Why
would you want to pass a relative path to a repository you are not in, and
this path being rooted elsewhere? Even if there is a need for this, it will
be the exceptional case, I think. Any other behaviour will force us to
convert paths before passing more often, don't you think?


2015-04-12 17:40 GMT+02:00 yarikoptic-private <notifications@github.com>:

> In datalad/support/annexrepo.py
> <https://github.com/datalad/datalad/pull/90#discussion_r28204456>:
>
> >
> > -    def __init__(self, path, url=None, runner=None):
> > +    Paths given to any of the class methods will be interpreted as relative to os.getcwd() in case this is currently beneath
>
> I still think we will burn ourself with this relative-to-getcwd() logic.
> E.g. now I see that in crawler I would need to guarantee I am in the root
> of the repo to manipulate using relative paths (or provide full paths all
> the time)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/datalad/datalad/pull/90/files#r28204456>.
>